eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php

```diff
@@ -61,13 +84,230 @@ protected function configure()
+        }
+
+        $iterations = ceil($count / $iterationCount);
+        if ($purge) {
+            $output->writeln('Purging index..');
+            $this->searchIndexer->purge($commit);
+            $message = "<info>Re-Creating index for {$count} content items across $iterations iteration(s)";
+        } else {
+            $message = "<info>Refreshing index for {$count} content items across $iterations iteration(s)";
+        }
+
+        $progress = new ProgressBar($output);
+
+        if ($iterations > 1 && ($processCount = (int) $input->getOption('processes')) > 1) {
+            // Don't run more processes then iterations
+            $processCount = $processCount > $iterations ? $iterations : $processCount;
+            $output->writeln($message . ", using $processCount parallel processes:</info>");
+            $progress->start($iterations);
+
+            return $this->runParallelProcess($progress, $stmt, (int) $processCount, (int) $iterationCount, $commit);
+        }
+
+        $output->writeln($message . ', using single process:</info>');
+
+        // if we only have one process, or less iterations to warrant running several, we index it all inline
+        $progress->start($iterations);
+        foreach ($this->fetchIteration($stmt, $iterationCount) as $contentIds) {
+            $this->searchIndexer->updateSearchIndex($contentIds, $commit);
+            $progress->advance(1);
+        }
+
+        $progress->finish();
+    }
+
+    private function runParallelProcess(ProgressBar $progress, Statement $stmt, $processCount, $iterationCount, $commit)
+    {
+        /**
+         * @var \Symfony\Component\Process\Process[]|null[]
+         */
+        $processes = array_fill(0, $processCount, null);
+        $generator = $this->fetchIteration($stmt, $iterationCount);
+        do {
+            foreach ($processes as $key => $process) {
+                if ($process !== null && $process->isRunning()) {
+                    continue;
+                }
+
+                if ($process !== null) {
+                    // One of the processes just finished, so we increment progress bar
+                    $progress->advance(1);
+                }
+
+                if (!$generator->valid()) {
+                    unset($processes[$key]);
+                    continue;
+                }
+
+                $processes[$key] = $this->getPhpProcess($generator->current(), $commit);
+                $processes[$key]->start();
+                $generator->next();
+            }
+
+            if (!empty($processes)) {
+                sleep(1);
+            }
+        } while (!empty($processes));
+
+        $progress->finish();
+    }
+
+    /**
+     * @param DateTime $since
+     * @param string $select Field to select, one of  'c.id' and 'count(c.id)'
+     *
+     * @return \Doctrine\DBAL\Driver\Statement
+     */
+    private function getStatementContentSince(DateTime $since, $select = 'c.id')
```

Comment:
This should be a part of Content Gateway (thus injecting either GW or in a cleaner fashion Persistence Content Handler to this command).
The reason for that is there's no easy way to track BC for database tables. There's Gateway that has a responsibility to consistently handle underlying raw database tables. This allows us to switch schema / storage engines in the future without worrying dependent things stop working.

Reply:
I know but same was the case in the Indexer you introduced, so until we have spi to get this, which is outside of the scope here, then adding private methods for this felt ok _(aka not worse then what we have)_

Reply:
> I know but same was the case in the Indexer you introduced, so until we have spi to get this, which is outside of the scope here, then adding private methods for this felt ok (aka not worse then what we have)

True, I guess I walked right into this one ;)
+1 then